File: app/src/main/java/com/sevtinge/cemiuiler/module/home/folder/FolderAutoClose.java
package com.sevtinge.cemiuiler.module.home.folder;

import android.view.View;

import com.sevtinge.cemiuiler.XposedInit;
import com.sevtinge.cemiuiler.module.base.BaseHook;

import de.robv.android.xposed.XposedHelpers;

public class FolderAutoClose extends BaseHook {

    @Override
    public void init() {
        findAndHookMethod("com.miui.home.launcher.Launcher", "launch", "com.miui.home.launcher.ShortcutInfo", View.class, new MethodHook() {
            @Override
            protected void after(MethodHookParam param) throws Throwable {
                if (!mPrefsMap.getBoolean("home_folder_auto_close")) return;
                boolean mHasLaunchedAppFromFolder = XposedHelpers.getBooleanField(param.thisObject, "mHasLaunchedAppFromFolder");
                if (mHasLaunchedAppFromFolder) XposedHelpers.callMethod(param.thisObject, "closeFolder");
            }
        });

        findAndHookMethodSilently("com.miui.home.launcher.common.CloseFolderStateMachine", "onPause", new MethodHook() {
            @Override
            protected void before(final MethodHookParam param) throws Throwable {
                if (XposedInit.mPrefsMap.getBoolean("home_folder_auto_close")) param.setResult(null);
            }
        });
    }
}

File: app/src/main/java/com/sevtinge/cemiuiler/module/home/folder/FolderAutoClose.kt
package com.sevtinge.cemiuiler.module.home.folder

import android.view.View
import com.sevtinge.cemiuiler.module.base.BaseHook
import com.sevtinge.cemiuiler.utils.woobox.callMethod
import com.sevtinge.cemiuiler.utils.woobox.getBooleanField
import com.sevtinge.cemiuiler.utils.woobox.hookAfterMethod

object FolderAutoClose : BaseHook() {
    override fun init() {
        "com.miui.home.launcher.Launcher".hookAfterMethod(
            "launch", "com.miui.home.launcher.ShortcutInfo", View::class.java
        ) {
            val mHasLaunchedAppFromFolder = it.thisObject.getBooleanField("mHasLaunchedAppFromFolder")
            if (mHasLaunchedAppFromFolder) it.thisObject.callMethod("closeFolder")
        }
    }
}